File: scripts/p4-audio.py
#!/usr/bin/env python3
"""Cama 6/8 original Aura Arcade + SFX alineados al manifiesto de captura."""
from __future__ import annotations

import json
import math
import random
import struct
import wave
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
RAW = ROOT / ".tmp-p4" / "raw"
OUT = ROOT / ".tmp-p4" / "audio"
SR = 48_000
RNG = random.Random(20260918)


def clamp(x: float, lo: float = -0.89, hi: float = 0.89) -> float:
    return lo if x < lo else hi if x > hi else x


def write_wav(path: Path, samples: list[float]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    peak = max((abs(s) for s in samples), default=1.0) or 1.0
    gain = min(1.0, 0.89 / peak)
    with wave.open(str(path), "w") as wav:
        wav.setnchannels(2)
        wav.setsampwidth(2)
        wav.setframerate(SR)
        frames = bytearray()
        for i, s in enumerate(samples):
            v = clamp(s * gain)
            # ligera apertura estéreo
            pan = 0.12 * math.sin(i / SR * 0.7)
            left = clamp(v * (1 - pan))
            right = clamp(v * (1 + pan))
            frames += struct.pack("<hh", int(left * 32767), int(right * 32767))
        wav.writeframes(frames)


def env(t: float, attack: float, dur: float) -> float:
    if t < 0 or t > dur:
        return 0.0
    if t < attack:
        return t / attack
    rel = (dur - t) / max(1e-4, dur - attack)
    return max(0.0, rel)


def tone(samples: list[float], at: float, freq: float, dur: float, amp: float, kind: str = "sine") -> None:
    n = len(samples)
    start = int(at * SR)
    length = int(dur * SR)
    for i in range(length):
        idx = start + i
        if idx >= n:
            break
        t = i / SR
        phase = 2 * math.pi * freq * t
        if kind == "tri":
            wave_v = 2 * abs(2 * ((t * freq) % 1) - 1) - 1
        elif kind == "saw":
            wave_v = 2 * ((t * freq) % 1) - 1
        else:
            wave_v = math.sin(phase)
        samples[idx] += wave_v * amp * env(t, 0.008, dur)


def noise(samples: list[float], at: float, dur: float, amp: float, hp: float = 400) -> None:
    start = int(at * SR)
    length = int(dur * SR)
    prev = 0.0
    for i in range(length):
        idx = start + i
        if idx >= n_safe(samples):
            break
        white = RNG.uniform(-1, 1)
        # highpass crude
        hp_v = white - prev
        prev = white
        t = i / SR
        samples[idx] += hp_v * amp * env(t, 0.004, dur) * min(1.0, hp / 800)


def n_safe(samples: list[float]) -> int:
    return len(samples)


def bed(seconds: float, bpm: float = 96.0) -> list[float]:
    n = int(seconds * SR) + SR
    samples = [0.0] * n
    eighth = 60.0 / (bpm * 1.5)  # 6/8: dotted-quarter = bpm
    bars = [
        (146.83, [293.66, 349.23, 440, 349.23, 293.66, 440]),
        (174.61, [349.23, 440, 523.25, 440, 349.23, 523.25]),
        (130.81, [261.63, 329.63, 392, 329.63, 261.63, 392]),
        (196.0, [392, 493.88, 587.33, 493.88, 392, 587.33]),
    ]
    t = 0.04
    bar_i = 0
    while t < seconds + 0.4:
        bass, notes = bars[bar_i % 4]
        for step in range(6):
            at = t + step * eighth
            shaker_amp = 0.018 if step in (0, 3) else 0.011
            noise(samples, at, 0.055, shaker_amp, 2800)
            if step in (0, 3):
                tone(samples, at, 92 if step == 0 else 72, 0.16, 0.05 if step == 0 else 0.034)
            tone(samples, at + 0.01, notes[step], 0.18 if step in (0, 3) else 0.12, 0.055 if step in (0, 3) else 0.036, "tri")
        tone(samples, t, bass, 0.42, 0.04)
        tone(samples, t + eighth * 3, bass * 1.5, 0.28, 0.026, "tri")
        t += eighth * 6
        bar_i += 1
    # aire nocturno
    for i in range(n):
        samples[i] += 0.008 * math.sin(2 * math.pi * 55 * i / SR) * (0.6 + 0.4 * math.sin(i / SR * 0.25))
    return samples[: int(seconds * SR)]


def mix_cue(samples: list[float], at: float, cue: str) -> None:
    if cue in ("humo-grab", "humo-start"):
        tone(samples, at, 196, 0.07, 0.09, "tri")
        tone(samples, at + 0.02, 392, 0.09, 0.07)
    elif cue == "humo-save":
        noise(samples, at, 0.12, 0.05, 800)
        tone(samples, at, 392, 0.08, 0.1, "tri")
        tone(samples, at + 0.05, 523, 0.12, 0.11, "tri")
        tone(samples, at + 0.1, 784, 0.16, 0.09)
    elif cue in ("pulso-hit", "pulso-now"):
        if cue == "pulso-now":
            tone(samples, at, 220, 0.05, 0.04)
            return
        tone(samples, at, 880, 0.18, 0.13, "tri")
        tone(samples, at + 0.04, 1320, 0.1, 0.06)
    elif cue == "radio-ok":
        noise(samples, at, 0.1, 0.04, 700)
        tone(samples, at, 392, 0.08, 0.1, "tri")
        tone(samples, at + 0.05, 523, 0.12, 0.11, "tri")
        tone(samples, at + 0.1, 784, 0.18, 0.1)
    elif cue == "radio-clue":
        tone(samples, at, 180, 0.12, 0.05, "saw")
    elif cue == "qr":
        tone(samples, at, 196, 0.4, 0.03)
        tone(samples, at + 0.12, 392, 0.5, 0.025, "tri")


def render(name: str, seconds: float, cues: list[tuple[float, str]]) -> Path:
    samples = bed(seconds, 96)
    for at, cue in cues:
        if 0 <= at < seconds - 0.05:
            mix_cue(samples, at, cue)
    # respiración antes del final
    fade = int(0.35 * SR)
    for i in range(fade):
        samples[-1 - i] *= i / fade
    out = OUT / f"{name}.wav"
    write_wav(out, samples)
    return out


def cues_from_timeline(timeline: list[dict]) -> list[tuple[float, str]]:
    return [(float(row["t"]), str(row["cue"])) for row in timeline]


def main() -> None:
    OUT.mkdir(parents=True, exist_ok=True)
    manifest = ROOT / ".tmp-p4" / "edit-timeline.json"
    if manifest.exists():
        data = json.loads(manifest.read_text(encoding="utf-8"))
        for item in data.get("mixes", []):
            path = render(item["name"], float(item["seconds"]), cues_from_timeline(item.get("cues", [])))
            print(path)
        return
    print(render("bed-18", 18, [(1.6, "humo-grab"), (4.6, "humo-save"), (8.4, "pulso-now"), (8.7, "pulso-hit"), (11.4, "radio-clue"), (12.2, "radio-ok"), (14.2, "qr")]))
    print(render("bed-12", 12, [(1.9, "humo-grab"), (5.4, "humo-save")]))
    print(render("bed-15", 15, [(1.4, "humo-grab"), (3.8, "humo-save"), (6.4, "pulso-now"), (6.7, "pulso-hit"), (10.0, "radio-clue"), (10.6, "radio-ok"), (12.2, "qr")]))


if __name__ == "__main__":
    main()

File: scripts/p4-audio.mjs
/**
 * Cama 6/8 original para P4 + SFX alineados al synth de HUMO/PULSO.
 * Receta: 96 BPM (negra con puntillo), compás 6/8, mismas funciones
 * Dm add9 / F maj7 / C add9 / G sus2 que `components/humo/humo-audio.ts`.
 * Licencia: composición procedural original Aura Arcade 2026. Sin samples de terceros.
 */
import { mkdir, writeFile } from 'node:fs/promises'
import path from 'node:path'

const SR = 48_000
const BPM = 96
const EIGHTH = 60 / BPM / 3
const OUT = path.join('.tmp-p4', 'audio')

const BARS = [
  { bass: 146.83, notes: [293.66, 349.23, 440, 349.23, 293.66, 440] },
  { bass: 174.61, notes: [349.23, 440, 523.25, 440, 349.23, 523.25] },
  { bass: 130.81, notes: [261.63, 329.63, 392, 329.63, 261.63, 392] },
  { bass: 196, notes: [392, 493.88, 587.33, 493.88, 392, 587.33] },
]

function mulberry(seed) {
  return () => {
    seed |= 0
    seed = (seed + 0x6d2b79f5) | 0
    let t = Math.imul(seed ^ (seed >>> 15), 1 | seed)
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

function clamp(v) {
  return v < -1 ? -1 : v > 1 ? 1 : v
}

function env(t, a, d) {
  if (t < 0) return 0
  if (t < a) return t / a
  const u = (t - a) / d
  if (u >= 1) return 0
  return Math.exp(-4.2 * u)
}

function tri(phase) {
  const x = phase - Math.floor(phase)
  return 1 - 4 * Math.abs(x - 0.5)
}

function mixAt(buf, i, v) {
  if (i >= 0 && i < buf.length) buf[i] += v
}

function renderTone(buf, start, dur, freq, gain, type) {
  const n = Math.floor(dur * SR)
  for (let i = 0; i < n; i++) {
    const t = i / SR
    const e = env(t, 0.008, dur)
    const ph = freq * t
    const w = type === 'sine' ? Math.sin(2 * Math.PI * ph) : type === 'saw' ? 2 * (ph - Math.floor(ph)) - 1 : tri(ph)
    mixAt(buf, start + i, w * e * gain)
  }
}

function renderNoise(buf, start, dur, gain, hp, rng) {
  const n = Math.floor(dur * SR)
  let prev = 0
  for (let i = 0; i < n; i++) {
    const t = i / SR
    const e = env(t, 0.004, dur)
    const white = rng() * 2 - 1
    const filtered = white - prev * hp
    prev = white
    mixAt(buf, start + i, filtered * e * gain)
  }
}

function pluck(buf, at, freq, dur, gain) {
  const start = Math.floor(at * SR)
  const n = Math.floor((dur + 0.04) * SR)
  for (let i = 0; i < n; i++) {
    const t = i / SR
    const e = env(t, 0.012, dur)
    const a = tri(freq * t)
    const b = Math.sin(2 * Math.PI * freq * 2.01 * t) * 0.35
    mixAt(buf, start + i, (a + b) * e * gain)
  }
}

function thump(buf, at, gain) {
  const start = Math.floor(at * SR)
  const n = Math.floor(0.18 * SR)
  for (let i = 0; i < n; i++) {
    const t = i / SR
    const freq = 92 * Math.pow(48 / 92, t / 0.14)
    const e = env(t, 0.006, 0.16)
    mixAt(buf, start + i, Math.sin(2 * Math.PI * freq * t) * e * gain)
  }
}

function shaker(buf, at, gain, rng) {
  const start = Math.floor(at * SR)
  const n = Math.floor(0.06 * SR)
  let lp = 0
  for (let i = 0; i < n; i++) {
    const t = i / SR
    const e = Math.exp(-t / 0.028)
    const white = rng() * 2 - 1
    lp = lp * 0.82 + white * 0.18
    const bp = white - lp
    mixAt(buf, start + i, bp * e * gain)
  }
}

function air(buf, rng) {
  let brown = 0
  for (let i = 0; i < buf.length; i++) {
    brown = clamp(brown + (rng() * 2 - 1) * 0.018)
    const hiss = (rng() * 2 - 1) * 0.012
    const night = Math.sin((i / SR) * 0.35) * 0.004
    buf[i] += brown * 0.045 + hiss * 0.02 + night
  }
}

function limit(buf, peakDb = -1.5) {
  let peak = 1e-9
  for (const s of buf) peak = Math.max(peak, Math.abs(s))
  const target = 10 ** (peakDb / 20)
  const g = peak > target ? target / peak : 1
  if (g !== 1) {
    for (let i = 0; i < buf.length; i++) buf[i] *= g
  }
  return { peak, gain: g, peakDb: 20 * Math.log10(peak * g) }
}

function writeWav(file, left, right = left) {
  const n = left.length
  const data = Buffer.alloc(44 + n * 4)
  data.write('RIFF', 0)
  data.writeUInt32LE(36 + n * 4, 4)
  data.write('WAVE', 8)
  data.write('fmt ', 12)
  data.writeUInt32LE(16, 16)
  data.writeUInt16LE(1, 20)
  data.writeUInt16LE(2, 22)
  data.writeUInt32LE(SR, 24)
  data.writeUInt32LE(SR * 4, 28)
  data.writeUInt16LE(4, 32)
  data.writeUInt16LE(16, 34)
  data.write('data', 36)
  data.writeUInt32LE(n * 4, 40)
  for (let i = 0; i < n; i++) {
    const l = Math.max(-1, Math.min(1, left[i] ?? 0))
    const r = Math.max(-1, Math.min(1, right[i] ?? l))
    data.writeInt16LE((l * 32767) | 0, 44 + i * 4)
    data.writeInt16LE((r * 32767) | 0, 46 + i * 4)
  }
  return writeFile(file, data)
}

function cue(name, fn) {
  const buf = new Float32Array(Math.floor(SR * 0.6))
  fn(buf)
  limit(buf, -3)
  return { name, buf }
}

await mkdir(OUT, { recursive: true })

const seconds = 24
const bed = new Float32Array(Math.floor(SR * seconds))
const rng = mulberry(0xa11a2026)
air(bed, rng)

let t = 0.06
let bar = 0
while (t < seconds - 1.4) {
  const chord = BARS[bar % BARS.length]
  for (let step = 0; step < 6; step++) {
    const at = t + step * EIGHTH
    shaker(bed, at, step === 0 || step === 3 ? 0.11 : 0.07, rng)
    if (step === 0 || step === 3) thump(bed, at, step === 0 ? 0.38 : 0.26)
    pluck(bed, at + 0.008, chord.notes[step], step === 0 || step === 3 ? 0.2 : 0.14, step === 0 || step === 3 ? 0.22 : 0.14)
  }
  pluck(bed, t, chord.bass, 0.44, 0.16)
  pluck(bed, t + EIGHTH * 3, chord.bass * 1.5, 0.3, 0.1)
  t += EIGHTH * 6
  bar += 1
}

const bedStats = limit(bed, -14)
await writeWav(path.join(OUT, 'bed-68-96bpm.wav'), bed)

const cues = [
  cue('grab', (b) => {
    renderTone(b, 0, 0.07, 196, 0.45, 'tri')
    renderTone(b, Math.floor(0.02 * SR), 0.09, 392, 0.32, 'sine')
  }),
  cue('tick', (b) => renderTone(b, 0, 0.04, 220, 0.22, 'sine')),
  cue('save', (b) => {
    const rng2 = mulberry(7)
    renderNoise(b, 0, 0.12, 0.28, 0.6, rng2)
    renderTone(b, 0, 0.08, 392, 0.42, 'tri')
    renderTone(b, Math.floor(0.05 * SR), 0.12, 523, 0.4, 'tri')
    renderTone(b, Math.floor(0.1 * SR), 0.16, 784, 0.32, 'sine')
  }),
  cue('perfect', (b) => {
    renderTone(b, 0, 0.18, 880, 0.42, 'tri')
    renderTone(b, Math.floor(0.04 * SR), 0.14, 1320, 0.18, 'sine')
  }),
  cue('si', (b) => {
    renderTone(b, 0, 0.08, 392, 0.38, 'tri')
    renderTone(b, Math.floor(0.05 * SR), 0.12, 523, 0.36, 'tri')
    renderTone(b, Math.floor(0.1 * SR), 0.18, 784, 0.3, 'sine')
  }),
  cue('fire', (b) => {
    const rng2 = mulberry(11)
    renderNoise(b, 0, 0.35, 0.22, 0.82, rng2)
    renderTone(b, 0, 0.4, 70, 0.28, 'sine')
  }),
]

for (const c of cues) {
  await writeWav(path.join(OUT, `${c.name}.wav`), c.buf)
}

await writeFile(
  path.join(OUT, 'manifest.json'),
  JSON.stringify(
    {
      bpm: BPM,
      meter: '6/8',
      sampleRate: SR,
      bedPeakDb: bedStats.peakDb,
      source: 'scripts/p4-audio.mjs mirrors components/humo/humo-audio.ts MUSIC_BARS',
      license: 'original Aura Arcade 2026',
    },
    null,
    2,
  ),
)

console.log(JSON.stringify({ ok: true, dir: OUT, bars: bar, bedPeakDb: bedStats.peakDb }))
